source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,209 @@
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length{};
+    StringUtil::atoul(content_length->value().c_str(), length);
+    return length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding && transfer_encoding->value().find(
+                                   Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    std::string content_type_str{content_type->value().c_str()};
+    StringUtil::eraseFrom(content_type_str, ";");
+    StringUtil::trim(content_type_str);
+    return config_->contentTypeValues().find(content_type_str) !=
+           config_->contentTypeValues().end();
+  }
+
+  return true;
+}
+
+bool GzipFilter::isCacheControlAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  return cache_control ? !cache_control->value().find(
+                             Http::Headers::get().CacheControlValues.NoTransform.c_str())
+                       : true;
+}
+
+bool GzipFilter::isTransferEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return !(transfer_encoding && transfer_encoding->value().find(
+                                    Http::Headers::get().TransferEncodingValues.Gzip.c_str()));
+}
+
+bool GzipFilter::isLastModifiedAllowed(HeaderMap& headers) const {
+  return !(config_->disableOnLastModified() && headers.LastModified());
+}
+
+bool GzipFilter::isEtagAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* etag = headers.Etag();
+  if (config_->disableOnEtag()) {
+    return !etag;
+  }
+
+  // TODO(gsagula): tests needed / do it efficiently.
+  if (etag && !etag->value().find("W/")) {
```

Comment:
I think the W/" has to be the first 3 characters in the etag.  Just check those?

https://github.com/abseil/abseil-cpp/blob/master/absl/strings/match.h